ENIGMAsystem/SHELL/Graphics_Systems/Direct3D11/DX11vertex.cpp

```diff
@@ -16,23 +16,279 @@
 **/
 
 #include "Graphics_Systems/General/GSvertex_impl.h"
+#include "Graphics_Systems/General/GScolor_macros.h"
+
+#include "Bridges/General/DX11Context.h"
+
+#include <D3Dcompiler.h>
+
+namespace {
+const char* g_strVS =
+  "struct VertexInputType {\n"
+  "  float4 position : POSITION;\n"
+  "  float4 color : COLOR;\n"
+  "};\n"
+  "struct PixelInputType {\n"
+  "  float4 position : SV_POSITION;\n"
+  "  float4 color : COLOR;\n"
+  "};\n"
+  "PixelInputType VS(VertexInputType input) {\n"
+  "  PixelInputType output;\n"
+  "  output.position = input.position;\n"
+  "  output.color = input.color;\n"
+  "  return output;\n"
+  "}\n";
+
+const char* g_strPS =
+  "struct PixelInputType {\n"
+  "  float4 position : SV_POSITION;\n"
+  "  float4 color : COLOR;\n"
+  "};\n"
+  "float4 PS(PixelInputType input) : SV_TARGET {\n"
+  "  return input.color;\n"
+  "}\n";
+
+ID3D10Blob* pBlobVS = NULL;
+ID3D10Blob* pBlobPS = NULL;
+
+D3D_PRIMITIVE_TOPOLOGY primitive_types[] = {
+  static_cast<D3D_PRIMITIVE_TOPOLOGY>(0),
+  D3D11_PRIMITIVE_TOPOLOGY_POINTLIST,
+  D3D11_PRIMITIVE_TOPOLOGY_LINELIST,
+  D3D11_PRIMITIVE_TOPOLOGY_LINESTRIP,
+  D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST,
+  D3D11_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP,
+  static_cast<D3D_PRIMITIVE_TOPOLOGY>(0) // fans are deprecated
+};
+
+const char* semantic_names[] = { "POSITION", "COLOR", "NORMAL", "TEXCOORD", "BLENDWEIGHT",
+  "BLENDINDICES", "DEPTH", "TANGENT", "BINORMAL", "FOG", "SAMPLE" };
+
+DXGI_FORMAT dxgi_formats[] = {
+  DXGI_FORMAT_R32_FLOAT,
+  DXGI_FORMAT_R32G32_FLOAT,
+  DXGI_FORMAT_R32G32B32_FLOAT,
+  DXGI_FORMAT_R32G32B32A32_FLOAT,
+  DXGI_FORMAT_R8G8B8A8_UNORM,
+  DXGI_FORMAT_R8G8B8A8_UINT
+};
+size_t dxgi_format_sizes[] = {
+  sizeof(float) * 1,
+  sizeof(float) * 2,
+  sizeof(float) * 3,
+  sizeof(float) * 4,
+  sizeof(unsigned byte) * 4,
+  sizeof(unsigned byte) * 4
+};
+
+map<int, ID3D11Buffer*> vertexBufferPeers;
+map<int, ID3D11Buffer*> indexBufferPeers;
+}
 
 namespace enigma {
 
-void graphics_delete_vertex_buffer_peer(int buffer) {}
+void graphics_delete_vertex_buffer_peer(int buffer) {
+  vertexBufferPeers[buffer]->Release();
+  vertexBufferPeers.erase(buffer);
+}
+
+void graphics_delete_index_buffer_peer(int buffer) {
+  indexBufferPeers[buffer]->Release();
+  indexBufferPeers.erase(buffer);
+}
+
+void graphics_prepare_buffer(const int buffer, const bool isIndex) {
+  const bool dirty = isIndex ? indexBuffers[buffer]->dirty : vertexBuffers[buffer]->dirty;
+  const bool frozen = isIndex ? indexBuffers[buffer]->frozen : vertexBuffers[buffer]->frozen;
+
+  // if the contents of the buffer are dirty then we need to update our native "peer"
+  if (dirty) {
+    ID3D11Buffer* bufferPeer = NULL;
+    auto it = isIndex ? indexBufferPeers.find(buffer) : vertexBufferPeers.find(buffer);
+    size_t size = isIndex ? enigma_user::index_get_size(buffer) : enigma_user::vertex_get_size(buffer);
```

Comment:
Why does this warrant a separate, independent lookup? Why is the size not inherent to the buffer class? I'd extract a utility method that computes the index/vertex size from an `ID3D11Buffer`.

Reply:
I can't do that without first constructing the ID3D11Buffer, because the byte size is a constructor parameter. The function `vertex_get_buffer_size` should return the size of the vertex buffer in bytes.
https://docs.yoyogames.com/source/dadiospice/002_reference/shaders/primitive%20building/vertex_get_buffer_size.html
The function `vertex_get_number` returns the number of vertices in the vertex buffer.
https://docs.yoyogames.com/source/dadiospice/002_reference/shaders/primitive%20building/vertex_get_number.html
I accidentally misnamed the `vertex_get_buffer_size` function which I am dealing with separately in #1284.

Reply:
Also, you should read my comments a few lines down describing where I compare the cached size with the queried size of the existing ID3D11Buffer to determine if I need to destroy it and make a bigger buffer.
https://github.com/enigma-dev/enigma-dev/blob/c075251ed90f8b1c3c667535343d257a555425d2/ENIGMAsystem/SHELL/Graphics_Systems/Direct3D11/DX11vertex.cpp#L113